File: lib/Controllers/skill_controller.dart
import 'package:flutter/material.dart';
import 'package:protfolio_app/Models/skill.dart';
import 'package:protfolio_app/services/skill_service.dart';

class SkillController extends ChangeNotifier {
  //step1 ----> here I will Create the class's variables :
  final SkillService _service = SkillService();
  List<Skill> _skills = [];
  bool _isLoading = false;
  String? _errorMessage;

  //step2 ----> here I will Create a getter && setter  :
  List<Skill> get skills => _skills;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;

  //step3 ----> here I will Create a function to load skills <get all skills> :
  Future<void> loadSkills() async {
    try {
      _isLoading = true;
      _errorMessage = null;
      notifyListeners();

      _skills = await _service.getAllSkills();
    } catch (exception) {
      _errorMessage = exception.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step4 ----> here I will Create a function to add skill :
  Future<void> addSkill(Skill skill) async {
    try {
      _isLoading = true;
      _errorMessage = null;

      notifyListeners();

      final newSkill = await _service.addSkill(skill);
      _skills.add(newSkill);
    } catch (e) {
      _errorMessage = e.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step5 ----> here I will Create a function to update skill :
  Future<void> updateSkill(String name, Skill skill) async {
    try {
      _isLoading = true;
      _errorMessage = null;

      notifyListeners();

      final updatedSkill = await _service.updateSkill(name, skill);
      final index = _skills.indexWhere((s) => s.name == name);

      if (index != -1) {
        _skills[index] = updatedSkill;
      }
    } catch (exception) {
      _errorMessage = exception.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step6 ----> here I will Create a function to delete skill :
  Future<void> deleteSkill(int id) async {
    try {
      _isLoading = true;
      _errorMessage = null;

      notifyListeners();

      await _service.deleteSkill(id);

      _skills.removeWhere((skill) => skill.id == id);
    } catch (exception) {
      _errorMessage = exception.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }
}

File: Backend/src/main/java/com/apps/PortfolioApp/Controller/SkillController.java
package com.apps.PortfolioApp.Controller;

import com.apps.PortfolioApp.DTO.SkillDTO;
import com.apps.PortfolioApp.Service.SkillService;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@RequestMapping("/skills")
public class SkillController {

    //Step1 ----> Create an object from Skill Service class :
    private final SkillService skillService;

    //Step2 ----> Create a con:
    public SkillController(SkillService skillService) {
        this.skillService = skillService;
    }

    //step3 ----> create a function to save the skill to db:
    @PostMapping
    public SkillDTO insert(@RequestBody @Valid SkillDTO skillDTO) {
        return this.skillService.insertSkill(skillDTO);
    }

    //step4 ----> create a function to get all skills from db:
    @GetMapping
    public List<SkillDTO> getAll() {
        return this.skillService.getAllSkills();
    }

    //step5 ----> create a function to get the Specific skill by id from db:
    @GetMapping("/id/{id}")
    public SkillDTO getById(@PathVariable int id){
        return this.skillService.getSkillByID(id);
    }

    //step6 ----> create a function to get the Specific skill by skillName from db:
    @GetMapping("/skillName/{name}")
    public SkillDTO getBySkillName(@PathVariable String name) {
        return this.skillService.getByName(name);
    }

    //step7 ----> create a function to update the skill to db:
    @PutMapping("/{name}")
    public SkillDTO update(@PathVariable String name , @RequestBody @Valid SkillDTO skillDTO) {
        return this.skillService.updateSkill(name , skillDTO);
    }

    //step8----> create a function to delete the Specific skill by id from db:
    @DeleteMapping("/{id}")
    public void delete(@PathVariable int id) {
        this.skillService.deleteSkill(id);
    }

}